### ISSUE TITLE
Tests for user pagination

### ISSUE THREAD

**Issue by CodeProKid** [wp-graphql/wp-graphql#274]
This issue spawned from #113. Implementing this with cursor pagination would be a total beast, since the appropriate filters don't currently exist in core to do this easily. This issue is to write tests to show that pagination is currently working with the `offset` and `limit` args. Then when we introduce real cursor pagination for this query, we will have tests to make sure we aren't breaking backwards compatibility.

A good test case would probably be creating a bunch of users with the create user factory method, and then doing multiple queries with different `first` and `last` args to make sure the proper data is being returned.

**Comment by Albert221** [comment]
I'd like to work on this issue.

**Comment by Albert221** [comment]
I'm unable to setup the environment. I'm doing steps given in the README, but there's a `WPDb module is not configured` that keeps me from doing it.

So, I've cloned the repo into my `wp-content/plugins` directory, then ran `bin/install-wp-tests.sh`, but turns out I need a WP-CLI in my `$PATH`, so I installed it, I ran it and it has finished. Then, I wanted to run the `codecept`, but I had to `composer install` firstly. Running the command again gives me this error: 

![image](https://user-images.githubusercontent.com/4085280/47608693-f98fbe80-da31-11e8-8018-17dfd1e2c5fa.png)

And I've provided the password on `install-wp-tests` command.

I'd like to complete this simple issue, but setting up the environment is so complex and - IMO - not well explained.

**Comment by jasonbahl** [comment]
@Albert221 ya, documentation for testing isn't the best. I'll work on updating it. 

If you're running the tests outside of Docker, check the instructions here: https://github.com/wp-graphql/wp-graphql#local-environment-configuration-for-codeception-tests

You'll need to override the config.yml to make sure your db user/pass for your local environment are configured properly.

**Comment by let00** [comment]
@CodeProKid @jasonbahl @Albert221  I can take a look at this issue too if y'all don't mind.

**Comment by jasonbahl** [comment]
@let00 if I'm not mistaken, pagination for users doesn't work _at all_ at the moment. 

We can get some failing tests in place to start, and work toward getting it to work, but I believe when I first approached cursor pagination for users, WP_User_Query didn't have proper filters to make it easy to work with. 

I believe @epeli did some digging while working on Cursors for Posts, and he believes there _are_ indeed filters we can use to implement cursor pagination properly for users. 

Related: #113

**Comment by epeli** [comment]
Yeah, while (re)implemting the cursors for PostsConnection I created a generic CursorBuilder class that can be used to create the where-clauses for all connections in wpgraphql:

https://github.com/wp-graphql/wp-graphql/blob/c2991573f65e56c35065538710ec62bf37c82b86/src/Data/Cursor/CursorBuilder.php

Although at moment it is only used in the PostsConnection.

The trouble with Users is that WordPress does not expose proper where-clause filter but it can be simulated like this:

```php
<?php

if ( ! defined( 'ABSPATH' ) ) {
    exit;
}

/**
 * Add missing filters to WP_User_Query class.
 */
add_filter( 'pre_user_query', function ( $query ) {

    if ( ! $query->get( 'suppress_filters' ) ) {
        $query->set( 'suppress_filters', 0 );
    }

    if ( ! $query->get( 'suppress_filters' ) ) {

        /**
         * Filters the FROM clause of the query.
         *
         * Specifically for manipulating paging queries.
         **
         *
         * @param string $where The FROM clause of the query.
         * @param WP_User_Query $query The WP_User_Query instance (passed by reference).
         */
        $query->query_from = apply_filters_ref_array( 'users_from', array( $query->query_from, &$query ) );

        /**
         * Filters the WHERE clause of the query.
         *
         * Specifically for manipulating paging queries.
         **
         *
         * @param string $where The WHERE clause of the query.
         * @param WP_User_Query $query The WP_User_Query instance (passed by reference).
         */
        $query->query_where = apply_filters_ref_array( 'users_where', array( $query->query_where, &$query ) );

        /**
         * Filters the ORDER BY clause of the query.
         *
         *
         * @param string $orderby The ORDER BY clause of the query.
         * @param WP_User_Query $query The WP_User_Query instance (passed by reference).
         */
        $query->query_orderby = apply_filters_ref_array( 'users_orderby', array( $query->query_orderby, &$query ) );

        /**
         * Filters the LIMIT clause of the query.
         *
         *
         * @param string $limits The LIMIT clause of the query.
         * @param WP_User_Query $query The WP_User_Query instance (passed by reference).
         */
        $query->query_limit = apply_filters_ref_array( 'users_limits', array( $query->query_limit, &$query ) );

        /**
         * Filters the SELECT clause of the query.
         *
         *
         * @param string $fields The SELECT clause of the query.
         * @param WP_User_Query $this The WP_User_Query instance (passed by reference).
         */
        $query->query_fields = apply_filters_ref_array( 'users_fields', array( $query->query_fields, &$query ) );

    }

    return $query;

} );

/**
 * Hook into users_where filter
 */
add_filter( 'users_where', function ( $where, $query ) {
    // Dirty example.
    return $where . ' AND 1=1';
}, 10, 2 );
```

**Comment by jasonbahl** [comment]
closed by #906
